pkg/controllers/mcs/mcs_controller.go

```diff
@@ -0,0 +1,361 @@
+	"k8s.io/client-go/tools/record"
+	"k8s.io/klog/v2"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+	mcsv1alpha1 "sigs.k8s.io/mcs-api/pkg/apis/v1alpha1"
+
+	networkingv1alpha1 "github.com/karmada-io/karmada/pkg/apis/networking/v1alpha1"
+	"github.com/karmada-io/karmada/pkg/events"
+	"github.com/karmada-io/karmada/pkg/sharedcli/ratelimiterflag"
+	"github.com/karmada-io/karmada/pkg/util"
+)
+
+const (
+	// MultiClusterServiceControllerName MultiClusterServiceController name.
+	MultiClusterServiceControllerName = "multi-cluster-service-controller"
+
+	// MultiClusterServiceControllerFinalizer is added to MultiClusterService to ensure related Works are deleted
+	// before MultiClusterService itself is deleted.
+	MultiClusterServiceControllerFinalizer = "karmada.io/multi-cluster-service-controller"
+)
+
+// MultiClusterServiceController lists/watches the multiClusterService resource,
+// generates ServiceExport and propagate it to the source clusters,
+// and generates ServiceImport and propagate it to the destination clusters.
+// Note: This controller only focus on  multiClusterService of type CrossCluster.
+type MultiClusterServiceController struct {
+	client.Client
+	EventRecorder      record.EventRecorder
+	RateLimiterOptions ratelimiterflag.Options
+}
+
+// Reconcile performs a full reconciliation for the object referred to by the Request.
+func (c *MultiClusterServiceController) Reconcile(ctx context.Context, req controllerruntime.Request) (controllerruntime.Result, error) {
+	klog.V(4).Infof("Reconciling MultiClusterService %s.", req.NamespacedName.String())
+
+	mcs := &networkingv1alpha1.MultiClusterService{}
+	if err := c.Client.Get(ctx, req.NamespacedName, mcs); err != nil {
+		if apierrors.IsNotFound(err) {
+			return controllerruntime.Result{}, nil
+		}
+		return controllerruntime.Result{Requeue: true}, err
+	}
+	ownerSvc := &corev1.Service{}
+	err := c.Client.Get(ctx, types.NamespacedName{Name: mcs.Name, Namespace: mcs.Namespace}, ownerSvc)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return controllerruntime.Result{}, nil
+		}
+		return controllerruntime.Result{Requeue: true}, err
+	}
+	if err := utilerrors.NewAggregate([]error{
+		c.reconcileServiceExport(ctx, mcs, ownerSvc),
+		c.reconcileServiceImport(ctx, mcs, ownerSvc),
+	}); err != nil {
+		klog.Errorf("Failed to build ServiceExport/ServiceImport for MultiClusterService(%s), error: %v", req.NamespacedName.String(), err)
+		return controllerruntime.Result{}, err
+	}
+	if !mcs.DeletionTimestamp.IsZero() || !mcs.Spec.HasExposureTypeCrossCluster() {
```

Comment:
Hi! @chaunceyjiang Suppose there is a scenario like this:
1. an MCS is created by the user.
2. The corresponding Service is created by the user, which triggers the logic to create ServiceImport and ServiceExport in the controller.
3. The Service is deleted by the user.
4. The user executes the command to delete the MCS. Since the relevant Service cannot be found, the controller returns at [line65](https://github.com/karmada-io/karmada/pull/3769/files#diff-a65aa54aa4bb0587ebb2f3b2cf4b9e03113b4be9c8535ae2fb57212f5540cc2bR65) and **there is no way to remove the `Finalizer` of the MCS**.
Please take a look at the scenario to see if it is correct, thanks!

Reply:
> there is no way to remove the Finalizer of the MCS.

Yes, because the Finalizer is being removed in another controller.

However, there are several implementation methods currently available, and I cannot determine which one is more reasonable. So I will discuss with everyone at today's community meeting. I hope you can come too.

Reply:
Thanks, I'm willing to learn the implementation methods!